### ISSUE TITLE
Location evaluation problem

### ISSUE THREAD

**Issue by knightyxp** [jia-wan/GeneralizedLoss-Counting-Pytorch#7]
hey jia wan:
     I have a lot of questions about the above test code of the location. First of all, in the training process, we input Keypoints, which are the marked points. I tried the method you mentioned above, but the count is always 0. The input density map you mentioned above uses the threshold to filter low confidence points. However, to locate the points this is not realistic for your method because your method input is an image and the output is keypoints, not a density map. I get confused about the location evaluation of GL Loss /DM Count , I am urgent to solve this problem. looking forward to your reply. my Gmail is knightyxp@gmail.com.

my location code evaluation is as below:

![image](https://user-images.githubusercontent.com/48432506/158226013-b36784d8-c07d-4871-9f85-409f3dc70c82.png)
![image](https://user-images.githubusercontent.com/48432506/158226181-22792633-f446-4372-99e9-ca4dd9d402fd.png)

**Comment by jia-wan** [comment]
Hi @knightyxp,

1. Please note that the localization is post-processing, which means that I don't compute the location during training. In particular, we first train the network to predict density maps correctly with the proposed loss function. Then, we can evaluate the localization performance based on the local minimum in the predicted density map. 
2. Please make sure the predicted density map is correct. You can check the total count of the predicted density map and visualize the density map.
3. You may need to find a proper threshold based on your training data.

Best,
Jia

**Comment by knightyxp** [comment]
Hi jia-wan:


> “Then, a pixel is considered a candidate for a predicted location if its value is the local maximum in a 3×3 window centered on  the pixel. Finally, the candidates with density larger than0.05 are the final location predictions. ”
> 
Aboved is quoted from your paper, I was confused whether it is the local minimum or local maximum to find the predicted location 
    I have visualized the predict density map which is all right, but I get confused about how to apply the local maximum to perform location performance, as far as I know, I even tried the threshold to 0.01, but the total predict people count is far less than the ground truth. (the sum of the predicted map is still closed to the ground truth)
   I wonder know how to find an appropriate threshold to locate the people in the density map

**Comment by jia-wan** [comment]
Hi @knightyxp,

1. Local maximum is the predicted location.
2. How to apply local maximum? Please refer to https://github.com/jia-wan/GeneralizedLoss-Counting-Pytorch/issues/3#issuecomment-922414636
3. An appropriate threshold can be found based on the validation set. I treat it as a hyper-parameter and tune it on the validation set. 

Best,
Jia

**Comment by knightyxp** [comment]
Hi jia-wan:
    The counting performance is too sensitive to the hyper-parameter of the threshold. In SHA,  I tried the 0.02/0.015/0.01, if the threshold is too high, the average count is larger than the gt, vice versa. I think a general threshold for the density map is not appropriate, there should be a threshold map for the location evaluation. my Gmail knightyxp@gmail.com , I want to further discuss this problem with u, looking forward to your reply
